### DISCUSSION TITLE
How to get a GPIO change interrupt on the RP2040?

### DISCUSSION THREAD

**Discussion by robzed** [tabemann/zeptoforth#50] [Q&A] [answered]
Is there any existing support for this as ZeptoForth words - or do a need to set CPU registers manually?

In micropython it would be something like:

   pin.irq(trigger=Pin.IRQ_FALLING, handler=callback).

I think the C/C++ API is something like:

gpio_set_irq_enabled_with_callback(21, GPIO_IRQ_EDGE_RISE | GPIO_IRQ_EDGE_FALL, true, &gpio_callback);
gpio_set_irq_enabled(22, GPIO_IRQ_EDGE_RISE | GPIO_IRQ_EDGE_FALL, true);
void gpio_callback(uint gpio, uint32_t events) {
    if(gpio==22)DoSomething();
    if(gpio==21)DoSomethingElse('.');
}

**Comment by tabemann** [comment]
This would be something like (note that I have not tried this code):

```
gpio import
variable old-handler
13 constant io-irq
io-irq 16 + constant io-vector
io-vector interrupt::vector@ old-handler !
true 21 PROC0_INTE_GPIO_EDGE_HIGH! true 21 PROC0_INTE_GPIO_EDGE_LOW!
true 22 PROC0_INTE_GPIO_EDGE_HIGH! true 22 PROC0_INTE_GPIO_EDGE_LOW!
:noname
  old-handler @ execute
  22 PROC0_INTE_GPIO_EDGE_HIGH@ 22 PROC0_INTE_GPIO_EDGE_LOW@ or if
    do-something 22 INTR_GPIO_EDGE_HIGH! 22 INTR_GPIO_EDGE_LOW!
  then
  21 PROC0_INTE_GPIO_EDGE_HIGH@ 21 PROC0_INTE_GPIO_EDGE_LOW@ or if
    s" ." do-something-else 21 INTR_GPIO_EDGE_HIGH! 21 INTR_GPIO_EDGE_LOW!
  then
; io-vector interrupt::vector!
```

**Reply by robzed** [reply]
thanks for this ... I'll try it out.

**Comment by tabemann** [comment]
Note that if you are compiling to flash, you will want this instead:

```
gpio import
variable old-handler
13 constant io-irq
io-irq 16 + constant io-vector
: register-my-gpio-handler ( -- )
  io-vector interrupt::vector@ old-handler !
  true 21 PROC0_INTE_GPIO_EDGE_HIGH! true 21 PROC0_INTE_GPIO_EDGE_LOW!
  true 22 PROC0_INTE_GPIO_EDGE_HIGH! true 22 PROC0_INTE_GPIO_EDGE_LOW!
  [:
    old-handler @ execute
    22 PROC0_INTE_GPIO_EDGE_HIGH@ 22 PROC0_INTE_GPIO_EDGE_LOW@ or if
      do-something 22 INTR_GPIO_EDGE_HIGH! 22 INTR_GPIO_EDGE_LOW!
    then
    21 PROC0_INTE_GPIO_EDGE_HIGH@ 21 PROC0_INTE_GPIO_EDGE_LOW@ or if
      s" ." do-something-else 21 INTR_GPIO_EDGE_HIGH! 21 INTR_GPIO_EDGE_LOW!
    then
  ;] io-vector interrupt::vector!
;
```

**Reply by robzed** [reply]
Do we have to clear INTR to ensure the interrupt doesn't happen again and again?

In the above code, 
I think the     22 PROC0_INTE_GPIO_EDGE_HIGH@ 22 PROC0_INTE_GPIO_EDGE_LOW@ or if
should be     22 PROC0_INTS_GPIO_EDGE_HIGH@ 22 PROC0_INTS_GPIO_EDGE_LOW@ or if
or maybe      22 PROC0_INTR_GPIO_EDGE_HIGH@ 22 PROC0_INTR_GPIO_EDGE_LOW@ or if

**Reply by tabemann** [reply]
That is what `22 INTR_GPIO_EDGE_HIGH! 22 INTR_GPIO_EDGE_LOW!` and `21 INTR_GPIO_EDGE_HIGH! 21 INTR_GPIO_EDGE_LOW!` do ─ they are clear-on-write. (Note that this only applies to edge interrupts ─ level interrupts if I gather it right keep in firing until the level changes.)

**Comment by robzed** [comment]
> That is what `22 INTR_GPIO_EDGE_HIGH! 22 INTR_GPIO_EDGE_LOW!` and `21 INTR_GPIO_EDGE_HIGH! 21 INTR_GPIO_EDGE_LOW!` do ─ they are clear-on-write. (Note that this only applies to edge interrupts ─ level interrupts if I gather it right keep in firing until the level changes.)

Ah, yes, I missed that in the RP2040 docs.

**Comment by robzed** [comment]
> Note that if you are compiling to flash, you will want this instead:
> 
> ```
> gpio import
> variable old-handler
> 13 constant io-irq
> io-irq 16 + constant io-vector
> : register-my-gpio-handler ( -- )
>   io-vector interrupt::vector@ old-handler !
>   true 21 PROC0_INTE_GPIO_EDGE_HIGH! true 21 PROC0_INTE_GPIO_EDGE_LOW!
>   true 22 PROC0_INTE_GPIO_EDGE_HIGH! true 22 PROC0_INTE_GPIO_EDGE_LOW!
>   [:
>     old-handler @ execute
>     22 PROC0_INTE_GPIO_EDGE_HIGH@ 22 PROC0_INTE_GPIO_EDGE_LOW@ or if
>       do-something 22 INTR_GPIO_EDGE_HIGH! 22 INTR_GPIO_EDGE_LOW!
>     then
>     21 PROC0_INTE_GPIO_EDGE_HIGH@ 21 PROC0_INTE_GPIO_EDGE_LOW@ or if
>       s" ." do-something-else 21 INTR_GPIO_EDGE_HIGH! 21 INTR_GPIO_EDGE_LOW!
>     then
>   ;] io-vector interrupt::vector!
> ;
> ```

When you are **selecting** which GPIO interrupt has fired to dispatch it, you are reading the interrupt enable register. (INTE). Surely you want to ue the INTS (interrupt status) register? (Not sure if you could also read the edge interrupt for INTR are as well ... it's only lists WC not read).

This is what I meant be second part of my comment:
I think the 22 PROC0_INTE_GPIO_EDGE_HIGH@ 22 PROC0_INTE_GPIO_EDGE_LOW@ or if
should be 22 PROC0_INTS_GPIO_EDGE_HIGH@ 22 PROC0_INTS_GPIO_EDGE_LOW@ or if

**Reply by tabemann** [reply, marked as answer]
Good catch! That should read:

```
gpio import
variable old-handler
13 constant io-irq
io-irq 16 + constant io-vector
: register-my-gpio-handler ( -- )
  io-vector interrupt::vector@ old-handler !
  true 21 PROC0_INTE_GPIO_EDGE_HIGH! true 21 PROC0_INTE_GPIO_EDGE_LOW!
  true 22 PROC0_INTE_GPIO_EDGE_HIGH! true 22 PROC0_INTE_GPIO_EDGE_LOW!
  [:
    old-handler @ execute
    22 PROC0_INTS_GPIO_EDGE_HIGH@ 22 PROC0_INTS_GPIO_EDGE_LOW@ or if
      do-something 22 INTR_GPIO_EDGE_HIGH! 22 INTR_GPIO_EDGE_LOW!
    then
    21 PROC0_INTS_GPIO_EDGE_HIGH@ 21 PROC0_INTS_GPIO_EDGE_LOW@ or if
      s" ." do-something-else 21 INTR_GPIO_EDGE_HIGH! 21 INTR_GPIO_EDGE_LOW!
    then
  ;] io-vector interrupt::vector!
;
```

**Reply by robzed** [reply]
Notes for everyone else: you need to enable the GPIO interrupt in the NIC (the interrupt controller) for the interrupts to trigger.

Something like:
`
interrupts import
13 NVIC_ISER_SETENA!
`

or

`13 bit NVIC_ISER_Base !`

The NIC interrupt for GPIO, IO_IRQ_BANK0 is 13. NVIC_ISER_Base address is the Interrupt Set enable register, which is at address $e000e100.

**Reply by tabemann** [reply]
I would recommend the following (all the comments are for those who may read this discussion who are less familiar with the RP2040):

```
gpio import
interrupt import

\ The previous GPIO IRQ handler
variable old-handler

\ The GPIO IRQ and interrupt vector index
13 constant io-irq
io-irq 16 + constant io-vector

\ Put your interrupt priority here; note that these values are from 0 to 3, and lower values are higher priority
0 constant my-priority

: register-my-gpio-handler ( -- )

  \ Disable the GPIO IRQ
  io-irq NVIC_ICER_CLRENA!

  \ Save the previous GPIO IRQ handler
  io-vector vector@ old-handler !

  \ Enable the relevant GPIO edge interrupts
  true 21 PROC0_INTE_GPIO_EDGE_HIGH! true 21 PROC0_INTE_GPIO_EDGE_LOW!
  true 22 PROC0_INTE_GPIO_EDGE_HIGH! true 22 PROC0_INTE_GPIO_EDGE_LOW!

  \ Our GPIO IRQ handler
  [:

    \ Execute the previous GPIO IRQ handler
    old-handler @ execute

    \ Handle edge high or edge low on GPIO 22
    22 PROC0_INTS_GPIO_EDGE_HIGH@ 22 PROC0_INTS_GPIO_EDGE_LOW@ or if
      \ Do our thing
      do-something 

      \ Clear the edge high or edge low interrupts on GPIO 22
      22 INTR_GPIO_EDGE_HIGH! 22 INTR_GPIO_EDGE_LOW!
    then

    \ Handle edge high or edge low on GPIO 21
    21 PROC0_INTS_GPIO_EDGE_HIGH@ 21 PROC0_INTS_GPIO_EDGE_LOW@ or if
      \ Do something else
      s" ." do-something-else
      
      \ Clear the edge high or edge low interrupt on GPIO 21
      21 INTR_GPIO_EDGE_HIGH! 21 INTR_GPIO_EDGE_LOW!
    then

  ;] io-vector vector!
  
  \ Set our priority for the GPIO IRQ
  my-priority 6 lshift io-irq NVIC_IPR_IP!

  \ Enable the GPIO IRQ
  io-irq NVIC_ISER_SETENA!

;
```
